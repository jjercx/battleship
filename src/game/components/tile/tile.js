import React, { useState } from "react";
import styled from "styled-components";
import * as actions from "game/redux/game.actions";
import { connect } from "react-redux";
import greenToken from "./assets/token-green.png";
import redToken from "./assets/token-red.png";
import { getCell } from "game/redux/game.reducer";
import { EMPTY } from "app/constants/game";

const Container = styled.div`
  grid-column-start: ${props => props.col + 2};
  grid-row-start: ${props => props.row + 2};
  background-color: ${props => props.theme.green}22;
`;

const Token = styled.div`
  height: 100%;
  width: 100%;
  background-image: url(${props => (props.empty ? greenToken : redToken)});
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
`;

export const Tile = ({ row, col, tileTouch, cell }) => {
  const [touched, setTouched] = useState(false);

  const handleClick = () => {
    if (!touched) {
      tileTouch({ row, col });
      setTouched(true);
    }
  };

  return (
    <Container onClick={handleClick} row={row} col={col} data-testid="tile">
      {touched && <Token empty={cell === EMPTY} data-testid="token" />}
    </Container>
  );
};

const mapStateToProps = (state, { row, col }) => ({
  cell: getCell(state, row, col),
});

export default connect(mapStateToProps, actions)(Tile);
