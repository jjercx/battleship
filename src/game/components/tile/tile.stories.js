import React from "react";
import Tile from "game/components/tile";
import styled from "styled-components";

export default {
  title: "game/Tile",
  component: Tile,
};

const Container = styled.div`
  display: grid;
  width: 50px;
  height: 50px;
`;

export const Default = () => (
  <Container>
    <Tile row={0} col={0} />
  </Container>
);
