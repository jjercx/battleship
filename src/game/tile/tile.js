import styled from "styled-components";

const Tile = styled.div`
  grid-column-start: ${props => props.col};
  grid-row-start: ${props => props.row};
`;

export default Tile;
