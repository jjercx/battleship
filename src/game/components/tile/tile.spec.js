import React from "react";
import { Tile } from "./tile";
import renderer from "react-test-renderer";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

describe("<Tile />", () => {
  it("matches snapshot", () => {
    const tree = renderer.create(<Tile />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe("when its touched", () => {
    it("reacts if touched for the first time", async () => {
      const handleClick = jest.fn();
      const row = 0;
      const col = 0;

      render(<Tile tileTouch={handleClick} col={col} row={row} />);
      fireEvent.click(screen.getByTestId("tile"));
      // await waitFor(() => screen.getByTestId("token")); // FIXME: PROBLEM WITH WAIT FOR

      expect(handleClick).toBeCalledTimes(1);
      expect(handleClick).toHaveBeenCalledWith({ row, col });
    });
    it("does nothing if already touched", () => {});
  });
});
