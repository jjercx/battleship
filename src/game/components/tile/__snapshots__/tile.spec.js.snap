// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile /> matches snapshot 1`] = `
<div
  className="sc-AxjAm jrksqQ"
  data-testid="tile"
  onClick={[Function]}
/>
`;
